### PR TITLE
fix: add refetch blocks and transactions fix command

### PR DESCRIPTION
**Pull request by MonikaCat**
## Description

References #216 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: cmd/fix/blocks/blocks.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+		return fmt.Errorf("error while updating block %v: %s", blockHeight, err)
+	}
+
+	return nil
+}
+
+func refreshTxs(parseCtx *parse.Context, consensusModule *consensus.Module, block *tmctypes.ResultBlock) error {
+
+	for _, tx := range block.Block.Txs {
+		var txDetails *juno.Tx
```
Comment (huichiaotsou) [inline cmd/fix/blocks/blocks.go]: I think we can remove the variable declaration at line 106 because declaration is made at line 108

File: modules/consensus/utils.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// GetStartingHeight takes starting height value from the config file
+func (m *Module) GetStartingHeight() int64 {
+	startHeight := m.cfg.StartHeight
+	return startHeight
+}
+
+// UpdateTxs updates txs in database
+func (m *Module) HandleMessages(txDetails *types.Tx) error {
```
Comment (huichiaotsou) [inline modules/consensus/utils.go]: The comment needs updating :p

File: database/consensus.go [hunk 3]
```diff
@@ -184,3 +193,145 @@ func (db *Db) GetGenesis() (*types.Genesis, error) {
...
+
+// CheckIfBlockIsMissing checks if block is already stored in database
+func (db *Db) CheckIfBlockIsMissing(height int64) bool {
+	var block []dbtypes.BlockRow
+	stmt := `SELECT * FROM block WHERE height = $1`
+
+	err := db.Sqlx.Select(&block, stmt, height)
+
+	return err != nil
+}
```
Comment (huichiaotsou) [inline database/consensus.go]: we should check "len(block) == 0" instead of checking "error != nil", otherwise it would always return false (the err would still be nil even when the selected row is empty)

Comment (MonikaCat) [reply database/consensus.go]: Ok I added the checks back to the code 👍🏻

**Review by huichiaotsou — APPROVED** [review]
LGTM, tested locally

File: database/consensus.go [hunk 4]
```diff
@@ -184,3 +193,151 @@ func (db *Db) GetGenesis() (*types.Genesis, error) {
...
+	}
+	if len(block) != 0 {
+		return false
+	}
+
+	return true
+}
+
+// UpdateBlocksInDatabase updates given block in database
+func (db *Db) UpdateBlockInDatabase(block *tmctypes.ResultBlock, blockResults *tmctypes.ResultBlockResults) error {
```
Comment (rmontagnin) [inline database/consensus.go]: Why do we need to re-create all these methods that are already part of the `Database` interface? Can't we simply use `db.SaveBlock`, `db.SaveTransaction` and `db.SaveMessage` instead?

File: database/staking_delegations.go [hunk 5]
```diff
@@ -27,6 +29,20 @@ func (db *Db) SaveDelegations(delegations []types.Delegation) error {
...
+			if scanErr != nil {
+				return fmt.Errorf("error while scanning error: %s", scanErr)
+			}
+
+			// If error is different than delegation not found, we need to return it
+			if errorCode != codes.NotFound.String() {
+				return fmt.Errorf("error while searching for delegations %s", err)
+			}
+		}
+
```
Comment (rmontagnin) [inline database/staking_delegations.go]: Why is this necessary?

File: modules/consensus/config.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package consensus
+
+import "gopkg.in/yaml.v3"
+
+// Config contains the configuration about chain start height
+type Config struct {
+	StartHeight int64 `yaml:"start_height,omitempty"`
+}
```
Comment (rmontagnin) [inline modules/consensus/config.go]: It's not necessary to define a new structure for this thing honestly. We can simply get the config that is parsed when the `bdjuno` command is used and get the starting height from there: 

```go
// Get the config
startHeight := config.Cfg.ParserStartHeight
```

File: database/consensus.go (hunk 4 above)
Comment (MonikaCat) [reply database/consensus.go]: Sure, I updated the code!

File: modules/consensus/config.go (hunk 6 above)
Comment (MonikaCat) [reply modules/consensus/config.go]: Updated

File: database/staking_delegations.go (hunk 5 above)
Comment (MonikaCat) [reply database/staking_delegations.go]: I added this to handle the error that bdjuno throws when it parses old delegate type tx and can't find it

Comment (rmontagnin) [reply database/staking_delegations.go]: I don't think we should ignore this error honestly. If the validator is not stored, we can't ignore the fact that something is wrong. Something like this should never happen

Comment (MonikaCat) [reply database/staking_delegations.go]: Okay sure! I've tested it locally and I can't reproduce this error anymore so I have removed this part from the code

File: cmd/fix/blocks/blocks.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/forbole/bdjuno/v2/utils"
+	"github.com/forbole/juno/v2/cmd/parse"
+	junomessages "github.com/forbole/juno/v2/modules/messages"
+	"github.com/forbole/juno/v2/types/config"
+	"github.com/spf13/cobra"
+	tmctypes "github.com/tendermint/tendermint/rpc/core/types"
+)
+
+// blocksCmd returns a Cobra command that allows to fix missing blocks in database
+func blocksCmd(parseConfig *parse.Config) *cobra.Command {
```
Comment (rmontagnin) [inline cmd/fix/blocks/blocks.go:14]: I think there is an easier way to implement this by using the existing `Worker` structure: 
1. Get the parsing context as you do
2. Build a new `Worker` instance
3. For each height, call the `worker.Process(height)` method returning any error

This will make sure that we call everything needed even if we add new modules in the future and would allow us to remove all the code duplication added to store things inside the database. 

**Note**: to implement this the currently private `process` method must be made public inside Juno.

Comment (MonikaCat) [reply cmd/fix/blocks/blocks.go:14]: Great! I updated the code and exported `process` method in juno repository

File: database/staking_delegations.go [hunk 8]
```diff
@@ -56,7 +56,7 @@ INSERT INTO delegation (validator_address, delegator_address, amount, height) VA
 		// Get the validator consensus address
 		consAddr, err := db.GetValidatorConsensusAddress(delegation.ValidatorOperAddr)
 		if err != nil {
-			return fmt.Errorf("error while gettting validator consensus address: %s", err)
+			return err
```
Comment (rmontagnin) [inline database/staking_delegations.go]: Can we revert this?

File: modules/staking/utils_delegations.go [hunk 9]
```diff
@@ -49,7 +45,7 @@ func (m *Module) getValidatorDelegations(height int64, validator string) ([]type
 func (m *Module) getDelegatorDelegations(height int64, delegator string) ([]types.Delegation, error) {
 	responses, err := m.source.GetDelegatorDelegations(height, delegator)
 	if err != nil {
-		return nil, fmt.Errorf("error while getting delegator delegations: %s", err)
```
Comment (rmontagnin) [inline modules/staking/utils_delegations.go]: Can we revert this?

File: modules/staking/utils_delegations.go [hunk 10]
```diff
@@ -85,7 +81,7 @@ func (m *Module) refreshDelegatorDelegations(height int64, delegator string) err
 	if err != nil {
 		// Get the error code
 		var code string
-		_, scanErr := fmt.Sscanf(err.Error(), ErrDelegationNotFound, &code, &code)
```
Comment (rmontagnin) [inline modules/staking/utils_delegations.go]: Can we revert this as well?

File: modules/utils/errors.go [hunk 11]
```diff
@@ -4,6 +4,8 @@ import (
 	"github.com/rs/zerolog/log"
 )
 
+const ErrNotFound = "rpc error: code = %s desc = rpc error: code = %s desc ="
```
Comment (rmontagnin) [inline modules/utils/errors.go]: No need to put this here since it's not used anywhere else than the delegations

**Review by rmontagnin — APPROVED** [review]
